Signal interface and derived classes

dianarg opened the pull request:
Adds classes that will be used to support MSRIOGroup refactoring.  These classes will replace MSRSignal, CombinedSignal & friends, and parts of the MSR + MSREncode usage in MSRIOGroup.

Resolves #1038

dianarg committed:
Add MSRIO::add_read() method

- Add overload of read_batch() with no return vector; the memory
  location will be returned by add_read().
- Will be used by individual MSR-based signals inside of MSRIOGroup to
  include themselves in batch reads.

Change-Id: Ic6d8225395efb64a65f83e6e07d2a66c7be50d5e
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add MSRFieldSignal class

Change-Id: I25125809a60cc02436ab9ed8beac50a5f811ae76
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add DifferenceSignal class

- Added MockSignal for unit tests.

Change-Id: I91d46ba0b8ca0240e5a384070edaa888aa76134e
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Move test helper functions to own file

Change-Id: I7ad04c3686ee181856f198ae23eb153901bc77ab
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add Signal base class

- Will be used to implement MSR field and combined signals in
  MSRIOGroup.

Change-Id: I5f95651ac78d1755edda22b6d10c4bb1a117fafa
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add RawMSRSignal class

- RawMSRSignal is used to define signals that copy the raw bits of an
  MSR into a 64-bit double with no SI conversion.
- Added MockMSRIO for unit testing.

Change-Id: Ia2d48cd45a908a06d2acafcab438d12153f69e41
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add DerivativeSignal class

Change-Id: I5baed7da04c5211f09bab4ba37e81dc919c14386
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add GEOPM_DEBUG_ASSERT macro for logic errors

- Should be used only for pre/post conditions that must be true based
  on guarantees of internal logic.
- Should not be used to check conditions that can be affected by user
  input to public APIs.

Change-Id: I4931ff156d74c01cddfea5968ab0e70e8d39ab1a
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg commented on src/RawMSRSignal.cpp:
```diff
@@ -0,0 +1,102 @@
+        GEOPM_DEBUG_ASSERT(m_msrio != nullptr, "no valid MSRIO object.");
+
+        if (!m_is_batch_ready) {
+            m_data = m_msrio->add_read(m_cpu, m_offset);
```
One change we discussed was to have this data pointer not be exposed here and add a sample() method to MSRIO.

bgeltz commented on src/DerivativeSignal.cpp:
```diff
@@ -0,0 +1,148 @@
+        , m_sleep_time(sleep_time)
+    {
+        GEOPM_DEBUG_ASSERT(m_time_sig && m_y_sig,
+                           "underlying Signals cannot be null.");
```
Underlying signals...

dianarg replied:
Maybe something like "Signal pointers for time_sig and y_sig cannot be null" ?

bgeltz commented on src/DerivativeSignal.hpp:
```diff
@@ -0,0 +1,78 @@
+            DerivativeSignal(std::shared_ptr<Signal> time_sig,
+                             std::shared_ptr<Signal> y_sig,
+                             int read_loops, double sleep_time);
+            DerivativeSignal(const DerivativeSignal &other);
```
Is the copy constructor used anywhere?

dianarg replied:
It's used in clone()

bgeltz commented on src/DerivativeSignal.cpp:
```diff
@@ -0,0 +1,148 @@
+            double signal = m_y_sig->read();
+            result = compute_next(temp_history, num_fit, time, signal);
+            if (ii < M_NUM_SAMPLE_HISTORY) {
+                usleep(m_sleep_time);            }
```
Closing curly is in the wrong spot.

bgeltz commented on src/DifferenceSignal.cpp:
```diff
@@ -0,0 +1,87 @@
+        , m_is_batch_ready(false)
+    {
+        GEOPM_DEBUG_ASSERT(m_minuend && m_subtrahend,
+                           "underlying Signals cannot be null.");
```
Same capitalization issues as before

bgeltz commented on src/DifferenceSignal.hpp:
```diff
@@ -0,0 +1,62 @@
+        public:
+            DifferenceSignal(std::shared_ptr<Signal> minuend,
+                             std::shared_ptr<Signal> subtrahend);
+            DifferenceSignal(const DifferenceSignal &other);
```
Is the copy constructor used?

bgeltz commented on src/MSRFieldSignal.hpp:
```diff
@@ -0,0 +1,91 @@
+                           int end_bit,
+                           int function,
+                           double scalar);
+            MSRFieldSignal(const MSRFieldSignal &other);
```
?

bgeltz reviewed (changes requested):
Looks mostly good to me.

A few nits on capitalization and curly brace placement.

Do we need all those copy constructors for the various signal types?  I didn't see where they were used.

dianarg reviewed (commented):
The copy constructors are needed because when the MSRIOGroup implements read_signal(), it will need to copy a shared_ptr<Signal> (case class pointer) but we need to know which copy constructor to call.  Also, some of the Signal types have state so the copy constructors are non-trivial.

dianarg committed:
Address review comments

- Remove clone() method and make read() const.
- Better error message.
- Fixed brace indentation.

Change-Id: I7f7689a0e529736b947ccf6fe02683141f55009c
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>
